### PR TITLE
SwiftIntroduction Practice

### PR DESCRIPTION
**Pull request by Sagamore-creator**
Dar nevisai gerai zinau ka rasyti sitoje vietoje :D Numetu dali practice, kiek spejau. Vieta kur reikia pagalbos pazymejau su MARK'u

### CODE REVIEW

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift [hunk 1]
```diff
@@ -68,17 +102,23 @@ struct RunApp {
...
+        didSet {
+            print(oldValue)
+            print(currentSong)
+        }
     }
 }
 
-var runApp = RunApp(id: "1", startTime: Date())
-//print(runApp.elapsedTime())
+var runApp = RunApp(id: "1", startTime: Date(), currentSong: Song(title: "Tralala", authorName: "Author", isPlaying: true))
```
Comment (martynasNarijauskas) [inline SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift]: We didn't have enough time for our lecture, but there is nice rule of 3 or more parameters when calling a function. 
Don't you think that this code is more readable?

```suggestion
var runApp = RunApp(
    id: "1",
    startTime: Date(),
    currentSong:
        Song(
            title: "Tralala",
            authorName: "Author",
            isPlaying: true
        )
)
```

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift [hunk 2]
```diff
@@ -134,48 +174,66 @@ class ColorPickerVC {
...
     
     init() {
-        onPublish = { imgName in
-            self.images.append(imgName)
-            //print("new img appended, img count: \(self.images.count)")
+        onPublish = { [weak self] imgName in
+            if let safeSelf = self {
+                safeSelf.images.append(imgName)
+                print("new img appended, img count: \(safeSelf.images.count)")
+            }
```
Comment (martynasNarijauskas) [inline SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift]: Nice! My recommendation, use guards in these places, so there will be less nesting. We could use guard let as a check for self, and after this check self is safe to use
```suggestion
            guard let self = self else {
                return
            }
            safeSelf.images.append(imgName)
            print("new img appended, img count: \(safeSelf.images.count)")
```

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift [hunk 3]
```diff
@@ -228,49 +303,67 @@ var arrayOfItems: [Item] = [
...
 let movies: [String: Float] = ["Home Alone 1": 10, "Star wars": 5, "Harry Potter": 8, "Home Alone 4": 2]
 
-//let mappedMovies: [String]
+func transformValueToString(dict: [String: Float]) {
+//    if (dict.map { $0.value }) > 5 {
+//
+//    } else if (dict.map { $0.value }) <= 5 {
+//
+//    }
+}
```
Comment (martynasNarijauskas) [inline SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift]: You can use map function and already pass block of code to your map. As I see from your code, you try to wrap 2 maps under 1 function, while you can fit all the logic under your map. I will show you an example and try to do it once again:)
```suggestion
let mappedMovies: [String] = movies.map { movie in
//    if movie.. {
//        return movie name and evaluation
//    } else {
//        return movie name and evaluation
//    }
}
try to solve it using this approach, if something is not clear, you can comment under here:)
```

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift [hunk 4]
```diff
@@ -228,49 +303,67 @@ var arrayOfItems: [Item] = [
...
-    if $0 % 2 == 0 {
-        newArray.append($0)
-    }
-}
+//arrayOfIntegers.forEach {
+//    if $0 % 2 == 0 {
+//        newArray.append($0)
+//    }
+//}
+newArray.append(contentsOf: arrayOfIntegers.filter { ($0 % 2) == 0 })
```
Comment (martynasNarijauskas) [inline SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift]: as we know, that our newArray is empty, we can simply set it as new array
```suggestion
newArray = arrayOfIntegers.filter { ($0 % 2) == 0 }
```

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift [hunk 5]
```diff
@@ -228,49 +303,67 @@ var arrayOfItems: [Item] = [
...
+var filteredArrayOfRandomObjects: [Any] = []
 
-//arrayOfRandomObjects.forEach {
-//    print("\($0)")
-//}
+for optionalElement in arrayOfRandomObjects {
+    if let unwrappedElement = optionalElement {
+        filteredArrayOfRandomObjects.append(unwrappedElement)
+    }
+}
```
Comment (martynasNarijauskas) [inline SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift]: try to check slide no. 22, there is a much easier way to do it :)

**Review by martynasNarijauskas — COMMENTED** [review]
Good job! really clean and well-formatted code. Take your time and try to finish the whole playground :)

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift (hunk 5 above)
Comment (Sagamore-creator) [reply SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift]: let arrayWithNoNils = arrayOfRandomObjects.compactMap { $0 }

File: SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+//
+// var imageVC: ImagePickerViewController? = ImagePickerViewController()
+//
+// imageVC?.buttonPressed()
+//
+// imageVC = nil
+
+print("PRACTICE DELEGATES:")
+
+protocol NetworkDelegate: AnyObject {
```
Comment (Sagamore-creator) [inline SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift:483]: Is this correct?

Comment (martynasNarijauskas) [reply SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift:483]: Yes, nicely done!:)

**Pull request closed without merging**